linalg/batchlinalg.cpp

```diff
@@ -0,0 +1,391 @@
+// Copyright (c) 2010-2024, Lawrence Livermore National Security, LLC. Produced
+// at the Lawrence Livermore National Laboratory. All Rights reserved. See files
+// LICENSE and NOTICE for details. LLNL-CODE-806117.
+//
+// This file is part of the MFEM library. For more information and source code
+// availability visit https://mfem.org.
+//
+// MFEM is free software; you can redistribute it and/or modify it under the
+// terms of the BSD-3 license. We welcome feedback and contributions, see file
+// CONTRIBUTING.md for details.
+
+
+// Implementation of batchlinalg class
+
+#include "batchlinalg.hpp"
+#include "../general/forall.hpp"
+
+#if defined(MFEM_USE_UMPIRE)
+#define MFEM_RELEASE_MEM(o) o.GetMemory().DeleteDevice(false)
+#else
+#define MFEM_RELEASE_MEM(o)
+#endif
+
+#if defined(MFEM_USE_CUDA)
+#include <cublas_v2.h>
+#include <cusolverDn.h>
+#define MFEM_cu_or_hip(stub) cu##stub
+#define MFEM_Cu_or_Hip(stub) Cu##stub
+#define MFEM_CU_or_HIP(stub) CU##stub
+#define MFEM_CUDA_or_HIP(stub) CUDA##stub
+#elif defined(MFEM_USE_HIP)
+#include <hipblas/hipblas.h>
+#include <hip/hip_runtime.h>
+#include <hip/hip_runtime_api.h>
+#define MFEM_cu_or_hip(stub) hip##stub
+#define MFEM_Cu_or_Hip(stub) Hip##stub
+#define MFEM_CU_or_HIP(stub) HIP##stub
+#define MFEM_CUDA_or_HIP(stub) HIP##stub
+#else
+#define MFEM_cu_or_hip(stub)
+#define MFEM_Cu_or_Hip(stub)
+#define MFEM_CU_or_HIP(stub)
+#define MFEM_CUDA_or_HIP(stub)
+#endif
+
+namespace mfem
+{
+
+#if defined(MFEM_USE_CUDA_OR_HIP)
+const MFEM_cu_or_hip(blasHandle_t) & DeviceBlasHandle()
+{
+   static MFEM_cu_or_hip(blasHandle_t) handle = nullptr;
+   static bool init                               = true;
+
+   if (init)
+   {
+      init        = false;
+      auto status = MFEM_cu_or_hip(blasCreate)(&handle);
+      MFEM_VERIFY(status == MFEM_CU_or_HIP(BLAS_STATUS_SUCCESS),
+                  "Cannot initialize GPU BLAS");
+      atexit([]()
+      {
+         MFEM_cu_or_hip(blasDestroy)(handle);
+         handle = nullptr;
+      });
+   }
+   return handle;
```

Comment:
Let's change this to let a library user set the blas handle:

```suggestion
static MFEM_cu_or_hip(blasHandle_t) device_blas_handle = nullptr;

const MFEM_cu_or_hip(blasHandle_t) & DeviceBlasHandle()
{
   if (!device_blas_handle)
   {
      auto status = MFEM_cu_or_hip(blasCreate)(&device_blas_handle);
      MFEM_VERIFY(status == MFEM_CU_or_HIP(BLAS_STATUS_SUCCESS),
                  "Cannot initialize GPU BLAS");
      atexit([]()
      {
         MFEM_cu_or_hip(blasDestroy)(device_blas_handle);
         device_blas_handle = nullptr;
      });
   }
   return device_blas_handle;
```

Then just put an extern definition in header!